### ISSUE TITLE
Error in log after sending notification

### ISSUE THREAD

**Issue by Andries-Smit** [mendix/MxPushNotifications#16]
Notifications are sent, however there are errors in the log.
It looks like a XML message ire received not a JSON message, sot it will not parse. 
issue #12 is maybe related to to such an issue, too.
> DEBUG - GCMConnection: Received: `<message><gcm xmlns="google:mobile:data">&lt;data:gcm xmlns:data=&quot;google:mobile:data&quot;&gt;</gcm></message>`
> ERROR - GCMConnection: Error parsing JSON `<data:gcm xmlns:data="google:mobile:data">`
> ERROR - GCMConnection: (1/17) org.json.simple.parser.ParseException: null
> ERROR - GCMConnection: (2/17) 	at org.json.simple.parser.Yylex.yylex(Unknown Source)
> ERROR - GCMConnection: (3/17) 	at org.json.simple.parser.JSONParser.nextToken(Unknown Source)
> ERROR - GCMConnection: (4/17) 	at org.json.simple.parser.JSONParser.parse(Unknown Source)
> ERROR - GCMConnection: (5/17) 	at org.json.simple.parser.JSONParser.parse(Unknown Source)
> ERROR - GCMConnection: (6/17) 	at org.json.simple.parser.JSONParser.parse(Unknown Source)
> ERROR - GCMConnection: (7/17) 	at org.json.simple.JSONValue.parseWithException(Unknown Source)
> ERROR - GCMConnection: (8/17) 	at pushnotifications.implementation.gcm.GCMPacketListener.processPacket(GCMPacketListener.java:35)
> ERROR - GCMConnection: (9/17) 	at org.jivesoftware.smack.XMPPConnection$ListenerWrapper.notifyListener(XMPPConnection.java:1233)
> ERROR - GCMConnection: (10/17) 	at org.jivesoftware.smack.XMPPConnection$ListenerNotification.run(XMPPConnection.java:1143)
> ERROR - GCMConnection: (11/17) 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
> ERROR - GCMConnection: (12/17) 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
> ERROR - GCMConnection: (13/17) 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
> ERROR - GCMConnection: (14/17) 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
> ERROR - GCMConnection: (15/17) 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
> ERROR - GCMConnection: (16/17) 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
> ERROR - GCMConnection: (17/17) 	at java.lang.Thread.run(Thread.java:745)

**Comment by pimvdnoll** [comment]
same issue andries :)

**Comment by simo101** [comment]
Yes i have seen this error before. I know where the error is coming from. But i don't understand why it would send back XML and not JSON. It's in the method processPacket in GCMPackageListener.

**Comment by Andries-Smit** [comment]
It looks like Google has some particularities when sending JSON, XML message structures with JSON payloads.

```
<gcm xmlns:google:mobile:data>
    JSON payload
</gcm>
```

[https://developers.google.com/cloud-messaging/ccs#format](https://developers.google.com/cloud-messaging/ccs#format)

So it looks that we need to trip the XML tag before parsing the JSON.

**Comment by simo101** [comment]
It seems very odd that the message we receive back from the GSM is not in JSON format.

**Comment by Andries-Smit** [comment]
Reference to Mendix Support Ticket [465928]

**Comment by ismailhabib** [comment]
The issue is in the implementation of the extension provider in `GCMConnection.java`. Instead of this:

```
ProviderManager.addExtensionProvider(GCM_ELEMENT_NAME, GCM_NAMESPACE,
				new PacketExtensionProvider() {
					@Override
					public PacketExtension parseExtension(XmlPullParser parser)
							throws Exception {
						String json = parser.getText();
						return new GcmPacketExtension(json);
					}
				});
```

Replace `String json = parser.getText();` with `String json = parser.nextText();`.

I will include this fix into the development-1.0 branch which will be the candidate for the next release. Meanwhile, feel free to manually fix this issue for your own project.

**Comment by KevinVlaanderen** [comment]
This issue was solve with release 1.0.0
